File: docs/breez-sdk/snippets/python/src/htlcs.py
# pylint: disable=duplicate-code
import hashlib
import logging
from typing import cast
from breez_sdk_spark import (
    BreezSdk,
    ClaimHtlcPaymentRequest,
    ListPaymentsRequest,
    PaymentDetails,
    PaymentDetailsFilter,
    PaymentStatus,
    PaymentType,
    PrepareSendPaymentRequest,
    ReceivePaymentMethod,
    ReceivePaymentRequest,
    SendPaymentRequest,
    SendPaymentOptions,
    SparkHtlcOptions,
    SparkHtlcStatus,
)


async def send_htlc_payment(sdk: BreezSdk):
    # ANCHOR: send-htlc-payment
    payment_request = "<spark address>"
    amount_sats = 50_000
    prepare_request = PrepareSendPaymentRequest(
        payment_request=payment_request,
        amount=amount_sats,
        token_identifier=None,
        conversion_options=None,
        fee_policy=None,
    )
    prepare_response = await sdk.prepare_send_payment(request=prepare_request)

    # If the fees are acceptable, continue to create the HTLC Payment
    if hasattr(prepare_response.payment_method, "fee"):
        fee = prepare_response.payment_method.fee
        logging.debug(f"Fees: {fee} sats")

    preimage = "<32-byte unique preimage hex>"
    preimage_bytes = bytes.fromhex(preimage)
    payment_hash_bytes = hashlib.sha256(preimage_bytes).digest()
    payment_hash = payment_hash_bytes.hex()

    # Set the HTLC options
    options = SendPaymentOptions.SPARK_ADDRESS(
        htlc_options=SparkHtlcOptions(
            payment_hash=payment_hash, expiry_duration_secs=1000
        )
    )

    request = SendPaymentRequest(
        prepare_response=prepare_response, options=options
    )
    send_response = await sdk.send_payment(request=request)
    payment = send_response.payment
    # ANCHOR_END: send-htlc-payment


async def receive_hodl_invoice_payment(sdk: BreezSdk):
    # ANCHOR: receive-hodl-invoice-payment
    preimage = "<32-byte unique preimage hex>"
    preimage_bytes = bytes.fromhex(preimage)
    payment_hash_bytes = hashlib.sha256(preimage_bytes).digest()
    payment_hash = payment_hash_bytes.hex()

    response = await sdk.receive_payment(
        request=ReceivePaymentRequest(
            payment_method=ReceivePaymentMethod.BOLT11_INVOICE(
                description="HODL invoice",
                amount_sats=50_000,
                expiry_secs=None,
                payment_hash=payment_hash,
            )
        )
    )

    invoice = response.payment_request
    logging.debug(f"HODL invoice: {invoice}")
    # ANCHOR_END: receive-hodl-invoice-payment


async def list_claimable_htlc_payments(sdk: BreezSdk):
    # ANCHOR: list-claimable-htlc-payments
    request = ListPaymentsRequest(
        type_filter=[PaymentType.RECEIVE],
        status_filter=[PaymentStatus.PENDING],
        payment_details_filter=[
            cast(PaymentDetailsFilter, PaymentDetailsFilter.SPARK(
                htlc_status=[SparkHtlcStatus.WAITING_FOR_PREIMAGE],
                conversion_refund_needed=None
            )),
            cast(PaymentDetailsFilter, PaymentDetailsFilter.LIGHTNING(
                htlc_status=[SparkHtlcStatus.WAITING_FOR_PREIMAGE],
            )),
        ],
    )

    response = await sdk.list_payments(request=request)
    payments = response.payments

    for payment in payments:
        if isinstance(payment.details, PaymentDetails.SPARK):
            if payment.details.htlc_details is not None:
                logging.debug(f"Spark HTLC expiry time: {payment.details.htlc_details.expiry_time}")
        elif isinstance(payment.details, PaymentDetails.LIGHTNING):
            expiry = payment.details.htlc_details.expiry_time
            logging.debug(f"Lightning HTLC expiry time: {expiry}")
    # ANCHOR_END: list-claimable-htlc-payments


async def claim_htlc_payment(sdk: BreezSdk):
    # ANCHOR: claim-htlc-payment
    preimage = "<preimage hex>"
    response = await sdk.claim_htlc_payment(
        request=ClaimHtlcPaymentRequest(preimage=preimage)
    )
    payment = response.payment
    # ANCHOR_END: claim-htlc-payment

File: docs/breez-sdk/snippets/go/htlcs.go
package example

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"log"
	"math/big"

	"github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
)

func SendHtlcPayment(sdk *breez_sdk_spark.BreezSdk) (*breez_sdk_spark.Payment, error) {
	// ANCHOR: send-htlc-payment
	paymentRequest := "<spark address>"
	// Set the amount you wish to pay the receiver
	amountSats := new(big.Int).SetInt64(50_000)
	prepareRequest := breez_sdk_spark.PrepareSendPaymentRequest{
		PaymentRequest:    paymentRequest,
		Amount:            &amountSats,
		TokenIdentifier:   nil,
		ConversionOptions: nil,
		FeePolicy:         nil,
	}
	prepareResponse, err := sdk.PrepareSendPayment(prepareRequest)

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	// If the fees are acceptable, continue to create the HTLC Payment
	switch paymentMethod := prepareResponse.PaymentMethod.(type) {
	case breez_sdk_spark.SendPaymentMethodSparkAddress:
		fee := paymentMethod.Fee
		log.Printf("Fees: %v sats", fee)
	}

	preimage := "<32-byte unique preimage hex>"
	preimageBytes, err := hex.DecodeString(preimage)
	if err != nil {
		return nil, err
	}
	paymentHashBytes := sha256.Sum256(preimageBytes)
	paymentHash := hex.EncodeToString(paymentHashBytes[:])

	// Set the HTLC options
	htlcOptions := breez_sdk_spark.SparkHtlcOptions{
		PaymentHash:        paymentHash,
		ExpiryDurationSecs: 1000,
	}
	var options breez_sdk_spark.SendPaymentOptions = breez_sdk_spark.SendPaymentOptionsSparkAddress{
		HtlcOptions: &htlcOptions,
	}

	request := breez_sdk_spark.SendPaymentRequest{
		PrepareResponse: prepareResponse,
		Options:         &options,
	}
	sendResponse, err := sdk.SendPayment(request)

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	payment := sendResponse.Payment
	// ANCHOR_END: send-htlc-payment
	return &payment, nil
}

func ReceiveHodlInvoicePayment(sdk *breez_sdk_spark.BreezSdk) error {
	// ANCHOR: receive-hodl-invoice-payment
	preimage := "<32-byte unique preimage hex>"
	preimageBytes, err := hex.DecodeString(preimage)
	if err != nil {
		return err
	}
	paymentHashBytes := sha256.Sum256(preimageBytes)
	paymentHash := hex.EncodeToString(paymentHashBytes[:])

	amountSats := uint64(50_000)
	response, err := sdk.ReceivePayment(breez_sdk_spark.ReceivePaymentRequest{
		PaymentMethod: breez_sdk_spark.ReceivePaymentMethodBolt11Invoice{
			Description: "HODL invoice",
			AmountSats:  &amountSats,
			ExpirySecs:  nil,
			PaymentHash: &paymentHash,
		},
	})

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return err
	}

	invoice := response.PaymentRequest
	log.Printf("HODL invoice: %v", invoice)
	// ANCHOR_END: receive-hodl-invoice-payment
	return nil
}

func ListClaimableHtlcPayments(sdk *breez_sdk_spark.BreezSdk) (*[]breez_sdk_spark.Payment, error) {
	// ANCHOR: list-claimable-htlc-payments
	typeFilter := []breez_sdk_spark.PaymentType{
		breez_sdk_spark.PaymentTypeReceive,
	}
	statusFilter := []breez_sdk_spark.PaymentStatus{
		breez_sdk_spark.PaymentStatusPending,
	}
	paymentDetailsFilter := []breez_sdk_spark.PaymentDetailsFilter{
		breez_sdk_spark.PaymentDetailsFilterSpark{
			HtlcStatus: &[]breez_sdk_spark.SparkHtlcStatus{
				breez_sdk_spark.SparkHtlcStatusWaitingForPreimage,
			},
		},
		breez_sdk_spark.PaymentDetailsFilterLightning{
			HtlcStatus: &[]breez_sdk_spark.SparkHtlcStatus{
				breez_sdk_spark.SparkHtlcStatusWaitingForPreimage,
			},
		},
	}

	request := breez_sdk_spark.ListPaymentsRequest{
		TypeFilter:            &typeFilter,
		StatusFilter:          &statusFilter,
		PaymentDetailsFilter:  &paymentDetailsFilter,
	}

	response, err := sdk.ListPayments(request)

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	payments := response.Payments

	for _, payment := range payments {
		if payment.Details != nil {
			switch details := (*payment.Details).(type) {
			case breez_sdk_spark.PaymentDetailsSpark:
				if details.HtlcDetails != nil {
					log.Printf("Spark HTLC expiry time: %v", details.HtlcDetails.ExpiryTime)
				}
			case breez_sdk_spark.PaymentDetailsLightning:
				log.Printf("Lightning HTLC expiry time: %v", details.HtlcDetails.ExpiryTime)
			}
		}
	}
	// ANCHOR_END: list-claimable-htlc-payments
	return &payments, nil
}

func ClaimHtlcPayment(sdk *breez_sdk_spark.BreezSdk) (*breez_sdk_spark.Payment, error) {
	// ANCHOR: claim-htlc-payment
	preimage := "<preimage hex>"
	request := breez_sdk_spark.ClaimHtlcPaymentRequest{
		Preimage: preimage,
	}
	response, err := sdk.ClaimHtlcPayment(request)

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	payment := response.Payment
	// ANCHOR_END: claim-htlc-payment
	return &payment, nil
}
